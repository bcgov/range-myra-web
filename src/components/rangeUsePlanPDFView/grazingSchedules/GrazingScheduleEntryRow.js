import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Table } from 'semantic-ui-react'
import * as utils from '../../../utils'
import { REFERENCE_KEY } from '../../../constants/variables'

class GrazingScheduleEntryRow extends Component {
  static propTypes = {
    entry: PropTypes.shape({}).isRequired,
    pasturesMap: PropTypes.shape({}).isRequired,
    references: PropTypes.shape({}).isRequired
  }

  render() {
    const { references, pasturesMap, entry } = this.props
    const livestockTypes = references[REFERENCE_KEY.LIVESTOCK_TYPE]
    const {
      id,
      pastureId,
      livestockTypeId,
      livestockCount,
      dateIn,
      dateOut,
      graceDays
    } = entry || {}

    const days = utils.calcDateDiff(dateOut, dateIn, false)
    const pasture = pasturesMap[pastureId]
    const pldPercent = pasture && pasture.pldPercent
    const pastureName = pasture && pasture.name
    const livestockType = livestockTypes.find(lt => lt.id === livestockTypeId)
    const livestockTypeName = livestockType && livestockType.name
    const auFactor = livestockType && livestockType.auFactor

    const totalAUMs = utils.calcTotalAUMs(livestockCount, days, auFactor)
    const pldAUMs = utils.roundTo1Decimal(
      utils.calcPldAUMs(totalAUMs, pldPercent)
    )
    const crownAUMs = utils.roundTo1Decimal(
      utils.calcCrownAUMs(totalAUMs, pldAUMs)
    )

    return (
      <Table.Row key={id}>
        <Table.Cell>{utils.handleNullValue(pastureName, false)}</Table.Cell>
        <Table.Cell>
          {utils.handleNullValue(livestockTypeName, false)}
        </Table.Cell>
        <Table.Cell collapsing>
          {utils.handleNullValue(livestockCount, false)}
        </Table.Cell>
        <Table.Cell>{utils.formatDateFromServer(dateIn, false)}</Table.Cell>
        <Table.Cell>{utils.formatDateFromServer(dateOut, false)}</Table.Cell>
        <Table.Cell collapsing>{utils.handleNullValue(days, false)}</Table.Cell>
        <Table.Cell collapsing>
          {utils.handleNullValue(graceDays || 0, false)}
        </Table.Cell>
        <Table.Cell collapsing>
          {utils.handleNullValue(pldAUMs, false)}
        </Table.Cell>
        <Table.Cell collapsing>
          {utils.handleNullValue(crownAUMs, false)}
        </Table.Cell>
      </Table.Row>
    )
  }
}

export default GrazingScheduleEntryRow
