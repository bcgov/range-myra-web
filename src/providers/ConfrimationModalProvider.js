import { Dialog, DialogActions, DialogContent, DialogContentText, DialogTitle } from '@material-ui/core';
import React, { createContext, useCallback, useContext, useRef, useState } from 'react';
import { PrimaryButton } from '../components/common';
const ConfirmationModal = createContext();

export function ConfirmationModalProvider({ children }) {
  const [state, setState] = useState({ isOpen: false });
  const fn = useRef();

  const confirm = useCallback(
    (data) => {
      return new Promise((resolve) => {
        setState({ ...data, isOpen: true });
        fn.current = (choice) => {
          resolve(choice);
          setState({ isOpen: false });
        };
      });
    },
    [setState],
  );

  return (
    <ConfirmationModal.Provider value={confirm}>
      {children}
      <Dialog
        open={state.isOpen}
        onClose={() => fn.current(false)}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
      >
        <DialogTitle id="alert-dialog-title">{state.titleText || 'Confirm'}</DialogTitle>
        <DialogContent>
          <DialogContentText id="alert-dialog-description">{state.contentText}</DialogContentText>
        </DialogContent>
        <DialogActions>
          <PrimaryButton inverted onClick={() => fn.current(false)}>
            {state.cancelText || 'Cancel'}
          </PrimaryButton>
          <PrimaryButton onClick={() => fn.current(true)} autoFocus>
            {state.confirmText || 'Confirm'}
          </PrimaryButton>
        </DialogActions>
      </Dialog>
    </ConfirmationModal.Provider>
  );
}

export default function useConfirm() {
  return useContext(ConfirmationModal);
}
