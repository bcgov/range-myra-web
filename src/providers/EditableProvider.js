import React, { useContext } from 'react'
import PropTypes from 'prop-types'

export const EditableContext = React.createContext({})

export const useEditable = () => useContext(EditableContext)

const EditableProvider = ({ children, editable }) => {
  return (
    <EditableContext.Provider value={editable}>
      {children}
    </EditableContext.Provider>
  )
}

EditableProvider.propTypes = {
  children: PropTypes.node
}

export default EditableProvider
