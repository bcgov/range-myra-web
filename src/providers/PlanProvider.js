import React, { useContext, useState, useEffect } from 'react'
import { normalize } from 'normalizr'
import * as API from '../api'
import { storePlan } from '../actions'
import * as schema from '../actionCreators/schema'
// import * as schema from '../schema'
import { getNetworkStatus } from '../utils/helper/network'
import { connect } from 'react-redux'
import { appendUsage } from '../utils'

/**
 * @typedef {Object} PlanContext
 * @property {string|number} currentPlanId ID of the current plan
 * @property {object|null} currentPlan Current plan
 * @property {boolean} isFetchingPlan Is the plan currently being fetched
 * @property {boolean} isSavingPlan Is the plan saving
 * @property {string|null} errorFetchingPlan Defined if there was an error fetching the plan
 * @property {(id?: number) => Promise<object>} fetchPlan Fetches by default the plan with id `currentPlanId`. Returns the plan, as well as sets `currentPlan`. If no network connectivity, fallbacks to local storage.
 * @property {(plan: object) => Promise<number>} savePlan Saves `plan` to either the remote backend or local storage, depending on network connectivity.
 */

/**
 * @type {React.Context<PlanContext>}
 */
const PlanContext = React.createContext()

/**
 * @returns {PlanContext}
 */
export const useCurrentPlan = () => useContext(PlanContext)

export const PlanProvider = ({ children, storePlan }) => {
  const [currentPlanId, setCurrentPlanId] = useState(null)
  const [currentPlan, setCurrentPlan] = useState(null)
  const [isFetchingPlan, setFetchingPlan] = useState(false)
  const [isSavingPlan, setSavingPlan] = useState(false)
  const [errorFetchingPlan, setErrorFetchingPlan] = useState(null)

  const fetchPlan = async (planId = currentPlanId) => {
    setFetchingPlan(true)

    try {
      const plan = await API.getPlan(planId)
      setCurrentPlan(appendUsage(plan))

      // TODO: remove redux
      const isOnline = await getNetworkStatus()
      if (isOnline) {
        // fetchRUP(planId)
        storePlan(normalize(plan, schema.plan))
      }

      return plan
    } catch (e) {
      setErrorFetchingPlan(e)
    } finally {
      setFetchingPlan(false)
    }
  }

  const savePlan = async plan => {
    setSavingPlan(true)

    const planId = await API.savePlan(plan)
    await fetchPlan(planId)

    setSavingPlan(false)

    return planId
  }

  useEffect(() => {
    if (currentPlanId !== null) {
      fetchPlan()
    }
  }, [currentPlanId])

  return (
    <PlanContext.Provider
      value={{
        setCurrentPlanId,
        currentPlan,
        isFetchingPlan,
        isSavingPlan,
        errorFetchingPlan,
        fetchPlan,
        savePlan
      }}>
      {children}
    </PlanContext.Provider>
  )
}

export default connect(
  null,
  { storePlan }
)(PlanProvider)
