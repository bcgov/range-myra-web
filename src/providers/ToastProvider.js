import React, { useContext, useState, useRef, useEffect } from 'react';
import uuid from 'uuid-v4';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import { Icon } from 'semantic-ui-react';

export const ToastContext = React.createContext({});

export const useToast = () => useContext(ToastContext);

const ToastProvider = ({ children }) => {
  const [toasts, setToasts] = useState([]);
  const toastsRef = useRef();

  useEffect(() => {
    toastsRef.current = toasts;
  }, [toasts]);

  const addToast = (message, status, config = {}) => {
    const { timeout = 3000, content } = config;
    const id = uuid();

    setToasts([
      ...toastsRef.current,
      {
        id,
        message,
        status,
        content,
      },
    ]);

    setTimeout(() => {
      removeToast(id);
    }, timeout);

    return id;
  };

  const successToast = (message, config) => addToast(message, 'success', config);
  const errorToast = (message, config) => addToast(message, 'error', config);
  const warningToast = (message, config) => addToast(message, 'warning', config);

  const removeToast = (id) => setToasts(toastsRef.current.filter((t) => t.id !== id));

  return (
    <ToastContext.Provider
      value={{
        addToast,
        successToast,
        errorToast,
        warningToast,
        removeToast,
      }}
    >
      <section className="toasts">
        {toasts.map((toast) => (
          <div key={toast.id} className="toast">
            <div
              className={classnames('toast__icon', {
                toast__icon__success: toast.status === 'success',
                toast__icon__error: toast.status === 'error',
                toast__icon__warning: toast.status === 'warning',
              })}
            >
              {toast.status === 'success' && <Icon name="check circle" size="large" />}
              {toast.status === 'error' && <Icon name="warning circle" size="large" />}
              {toast.status === 'warning' && <Icon name="warning" size="large" />}
            </div>
            <div className="toast__content">
              {toast.message}
              {toast.content}
            </div>
            <button className="toast__dismiss" onClick={() => removeToast(toast.id)}>
              <Icon name="times" size="small" />
            </button>
          </div>
        ))}
      </section>
      {children}
    </ToastContext.Provider>
  );
};

ToastProvider.propTypes = {
  user: PropTypes.shape({
    id: PropTypes.number.isRequired,
    email: PropTypes.string.isRequired,
    givenName: PropTypes.string.isRequired,
    familyName: PropTypes.string.isRequired,
    username: PropTypes.string.isRequired,
    roles: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  }),
  children: PropTypes.node,
};

export default ToastProvider;
