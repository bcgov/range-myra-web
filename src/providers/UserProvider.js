import React, { useContext } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getUser } from '../reducers/rootReducer';

export const UserContext = React.createContext({});

export const useUser = () => useContext(UserContext);

const UserProvider = ({ user, children }) => {
  return <UserContext.Provider value={user}>{children}</UserContext.Provider>;
};

UserProvider.propTypes = {
  user: PropTypes.shape({
    id: PropTypes.number.isRequired,
    email: PropTypes.string.isRequired,
    givenName: PropTypes.string.isRequired,
    familyName: PropTypes.string.isRequired,
    username: PropTypes.string.isRequired,
    roles: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  }),
  children: PropTypes.node,
};

// Just take user from redux store and make it accessible via UserContext for now
// Eventually handle retrieving and storing the user in this component w/o redux
const mapStateToProps = (state) => ({
  user: getUser(state),
});

export default connect(mapStateToProps)(UserProvider);
