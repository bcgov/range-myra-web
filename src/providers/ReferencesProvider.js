import React, { useContext, useEffect } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { getReferences } from '../reducers/rootReducer'
import { fetchReferences } from '../actionCreators'
import { useUser } from './UserProvider'

export const ReferencesContext = React.createContext({})

export const useReferences = () => useContext(ReferencesContext)

const ReferencesProvider = ({ references, fetchReferences, children }) => {
  const user = useUser()

  useEffect(() => {
    fetchReferences()
  }, [user])

  return (
    <ReferencesContext.Provider value={references}>
      {children}
    </ReferencesContext.Provider>
  )
}

ReferencesProvider.propTypes = {
  references: PropTypes.shape({}).isRequired,
  fetchReferences: PropTypes.func.isRequired,
  children: PropTypes.node
}

// Just take references from redux store and make it accessible via ReferencesContext for now
// Eventually handle retrieving and storing the references in this component w/o redux
const mapStateToProps = state => ({
  references: getReferences(state)
})

const mapDispatchToProps = {
  fetchReferences
}

export default connect(mapStateToProps, mapDispatchToProps)(ReferencesProvider)
