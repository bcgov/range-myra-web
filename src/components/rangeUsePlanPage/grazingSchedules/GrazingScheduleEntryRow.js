import React from 'react'
import PropTypes from 'prop-types'
import { Table, Dropdown, Icon } from 'semantic-ui-react'
import { Dropdown as FormikDropdown } from 'formik-semantic-ui'
import { connect } from 'formik'
import * as utils from '../../../utils'
import { useReferences } from '../../../providers/ReferencesProvider'
import { REFERENCE_KEY } from '../../../constants/variables'
import PermissionsField, { IfEditable } from '../../common/PermissionsField'
import { SCHEDULE } from '../../../constants/fields'
import DateInputField from '../../common/form/DateInputField'
import moment from 'moment'
import PasturesDropdown from './PasturesDropdown'

const GrazingScheduleEntryRow = ({
  entry,
  formik,
  namespace,
  onDelete,
  onCopy,
  schedule
}) => {
  const {
    pastureId,
    livestockTypeId,
    livestockCount,
    dateIn,
    dateOut,
    graceDays
  } = entry || {}

  const references = useReferences()
  const livestockTypes = references[REFERENCE_KEY.LIVESTOCK_TYPE]

  const livestockTypeOptions = livestockTypes.map(lt => {
    const { id, name } = lt || {}
    return {
      key: id,
      value: id,
      text: name
    }
  })

  const days = utils.calcDateDiff(dateOut, dateIn, false)
  const pasture = formik.values.pastures.find(p => p.id === pastureId)

  const pldPercent = pasture && pasture.pldPercent
  const livestockType = livestockTypes.find(lt => lt.id === livestockTypeId)
  const auFactor = livestockType && livestockType.auFactor

  const totalAUMs = utils.calcTotalAUMs(livestockCount, days, auFactor)
  const pldAUMs = utils.roundTo1Decimal(
    utils.calcPldAUMs(totalAUMs, pldPercent)
  )
  const crownAUMs = utils.roundTo1Decimal(
    utils.calcCrownAUMs(totalAUMs, pldAUMs)
  )

  const entryOptions = [
    { key: 'copy', text: 'Duplicate', onClick: onCopy },
    {
      key: 'delete',
      text: 'Delete',
      onClick: onDelete
    }
  ]

  const initialDate = moment()
    .set('year', schedule.year)
    .set('month', 0)
    .set('date', 1)

  const maxDate = moment()
    .set('year', schedule.year)
    .set('month', 11)
    .set('date', 31)

  return (
    <Table.Row>
      <Table.Cell>
        <PasturesDropdown
          name={`${namespace}.pastureId`}
          pastureId={pastureId}
          onChange={(e, { pasture }) => {
            formik.setFieldValue(
              `${namespace}.graceDays`,
              pasture.graceDays || 0
            )
          }}
        />
      </Table.Cell>
      <Table.Cell>
        <PermissionsField
          permission={SCHEDULE.TYPE}
          name={`${namespace}.livestockTypeId`}
          options={livestockTypeOptions}
          component={FormikDropdown}
          displayValue={
            livestockTypeOptions.find(o => o.value === livestockTypeId)
              ? livestockTypeOptions.find(o => o.value === livestockTypeId).text
              : ''
          }
          inputProps={{
            search: true,
            'aria-label': 'livestock type'
          }}
          fast
        />
      </Table.Cell>
      <Table.Cell collapsing>
        <PermissionsField
          permission={SCHEDULE.TYPE}
          name={`${namespace}.livestockCount`}
          displayValue={livestockCount}
          inputProps={{
            'aria-label': 'livestock count'
          }}
          fast
        />
      </Table.Cell>
      <Table.Cell collapsing>
        <PermissionsField
          permission={SCHEDULE.DATE_IN}
          name={`${namespace}.dateIn`}
          component={DateInputField}
          displayValue={moment(dateIn).format('MMM D')}
          dateFormat="MMM D YYYY"
          icon={null}
          initialDate={initialDate}
          minDate={initialDate}
          maxDate={maxDate}
          aria-label="date in"
        />
      </Table.Cell>
      <Table.Cell collapsing>
        <PermissionsField
          permission={SCHEDULE.DATE_OUT}
          name={`${namespace}.dateOut`}
          component={DateInputField}
          displayValue={moment(dateOut).format('MMM D')}
          dateFormat="MMM D YYYY"
          icon={null}
          initialDate={initialDate}
          minDate={initialDate}
          maxDate={maxDate}
          aria-label="date out"
        />
      </Table.Cell>
      <Table.Cell collapsing>{utils.handleNullValue(days, false)}</Table.Cell>
      <Table.Cell collapsing>
        <PermissionsField
          permission={SCHEDULE.GRACE_DAYS}
          name={`${namespace}.graceDays`}
          displayValue={graceDays || (pasture && pasture.graceDays) || 0}
          inputProps={{
            type: 'number',
            'aria-label': 'grace days'
          }}
          fieldProps={{
            onBlur: e => {
              if (e.target.value.trim() === '') {
                formik.setFieldValue(`${namespace}.graceDays`, 0)
              }
            }
          }}
          fast
        />
      </Table.Cell>
      <Table.Cell collapsing>
        {utils.handleNullValue(pldAUMs, false)}
      </Table.Cell>
      <Table.Cell collapsing>
        {utils.handleNullValue(crownAUMs, false)}
      </Table.Cell>
      <IfEditable permission={SCHEDULE.TYPE}>
        <Table.Cell collapsing textAlign="center">
          <Dropdown
            trigger={<Icon name="ellipsis vertical" />}
            options={entryOptions}
            icon={null}
            pointing="right"
          />
        </Table.Cell>
      </IfEditable>
    </Table.Row>
  )
}

GrazingScheduleEntryRow.propTypes = {
  entry: PropTypes.object.isRequired,
  formik: PropTypes.object.isRequired,
  namespace: PropTypes.string.isRequired,
  onDelete: PropTypes.func.isRequired,
  onCopy: PropTypes.func.isRequired
}

export default connect(GrazingScheduleEntryRow)
