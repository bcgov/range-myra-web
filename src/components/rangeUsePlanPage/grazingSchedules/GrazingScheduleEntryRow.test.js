import React from 'react'
import uuid from 'uuid-v4'
import { Formik } from 'formik'
import { render } from '../../../tests/helpers/test-utils'
import GrazingScheduleEntryRow from './GrazingScheduleEntryRow'
import { UserContext } from '../../../providers/UserProvider'

const schedule = {
  id: 1,
  year: 2022,
  grazingScheduleEntries: [
    {
      pastureId: 1,
      dateIn: new Date().toISOString(),
      dateOut: new Date().toISOString(),
      graceDays: 4,
      livestockCount: 10,
      id: uuid()
    }
  ]
}

const pastures = [{ id: 1, name: 'Pasture 1' }, { id: 2, name: 'Pasture 2' }]

const namespace = 'schedule.grazingScheduleEntries.0'

describe('Grazing Schedule Box', () => {
  it('Displays each grazing schedule entry', () => {
    const { container } = render(
      <Formik
        initialValues={{ schedule, pastures }}
        render={({ values }) => (
          <table>
            <tbody>
              <GrazingScheduleEntryRow
                schedule={values.schedule}
                entry={values.schedule.grazingScheduleEntries[0]}
                namespace={namespace}
                onCopy={jest.fn()}
                onDelete={jest.fn()}
              />
            </tbody>
          </table>
        )}
      />
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('shows an input for each schedule entry field', () => {
    const { queryByLabelText, getByLabelText } = render(
      <Formik
        initialValues={{ schedule, pastures }}
        render={({ values }) => (
          <table>
            <tbody>
              <GrazingScheduleEntryRow
                schedule={values.schedule}
                entry={values.schedule.grazingScheduleEntries[0]}
                namespace="schedule.grazingScheduleEntries.0"
                onCopy={jest.fn()}
                onDelete={jest.fn()}
              />
            </tbody>
          </table>
        )}
      />
    )

    expect(getByLabelText('pasture')).toBeInTheDocument()
    expect(getByLabelText('livestock type')).toBeInTheDocument()
    expect(getByLabelText('livestock count')).toBeInTheDocument()
    expect(getByLabelText('date in')).toBeInTheDocument()
    expect(getByLabelText('date out')).toBeInTheDocument()
    expect(queryByLabelText('grace days')).toBeNull()
  })

  it('only shows the grace days input for range officers', () => {
    const user = {
      roles: ['myra_range_officer']
    }

    const { getByLabelText, queryByLabelText } = render(
      <UserContext.Provider value={user}>
        <Formik
          initialValues={{ schedule, pastures }}
          render={({ values }) => (
            <table>
              <tbody>
                <GrazingScheduleEntryRow
                  schedule={values.schedule}
                  entry={values.schedule.grazingScheduleEntries[0]}
                  namespace="schedule.grazingScheduleEntries.0"
                  onCopy={jest.fn()}
                  onDelete={jest.fn()}
                />
              </tbody>
            </table>
          )}
        />
      </UserContext.Provider>
    )

    expect(getByLabelText('grace days')).toBeInTheDocument()
    expect(queryByLabelText('pasture')).toBeNull()
    expect(queryByLabelText('livestock type')).toBeNull()
    expect(queryByLabelText('livestock count')).toBeNull()
    expect(queryByLabelText('date in')).toBeNull()
    expect(queryByLabelText('date out')).toBeNull()
  })
})
