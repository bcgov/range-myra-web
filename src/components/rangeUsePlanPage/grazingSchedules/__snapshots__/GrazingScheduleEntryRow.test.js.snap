// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grazing Schedule Box Displays each grazing schedule entry 1`] = `
<table>
  <tbody>
    <tr
      class=""
    >
      <td
        class=""
      >
        <div
          class="field"
        >
          <div
            aria-expanded="false"
            aria-label="pasture"
            class="ui fluid search selection dropdown"
            id="field_dropdown_schedule.grazingScheduleEntries.0.pastureId"
            name="schedule.grazingScheduleEntries.0.pastureId"
            role="combobox"
          >
            <input
              aria-autocomplete="list"
              autocomplete="off"
              class="search"
              tabindex="0"
              type="text"
              value=""
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              class="text"
              role="alert"
            >
              Pasture 1
            </div>
            <i
              aria-hidden="true"
              class="dropdown icon"
            />
            <div
              class="menu transition"
              role="listbox"
            >
              <div
                aria-checked="true"
                aria-selected="true"
                class="active selected item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  Pasture 1
                </span>
              </div>
              <div
                aria-checked="false"
                aria-selected="false"
                class="item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  Pasture 2
                </span>
              </div>
            </div>
          </div>
        </div>
      </td>
      <td
        class=""
      >
        <div
          class="field"
        >
          <div
            aria-expanded="false"
            aria-label="livestock type"
            class="ui fluid search selection dropdown"
            id="field_dropdown_schedule.grazingScheduleEntries.0.livestockTypeId"
            name="schedule.grazingScheduleEntries.0.livestockTypeId"
            role="combobox"
          >
            <input
              aria-autocomplete="list"
              autocomplete="off"
              class="search"
              tabindex="0"
              type="text"
              value=""
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              class="text"
              role="alert"
            />
            <i
              aria-hidden="true"
              class="dropdown icon"
            />
            <div
              class="menu transition"
              role="listbox"
            >
              <div
                aria-checked="false"
                aria-selected="true"
                class="selected item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  Cow with Calf
                </span>
              </div>
              <div
                aria-checked="false"
                aria-selected="false"
                class="item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  Bull
                </span>
              </div>
              <div
                aria-checked="false"
                aria-selected="false"
                class="item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  Yearling
                </span>
              </div>
              <div
                aria-checked="false"
                aria-selected="false"
                class="item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  Horse
                </span>
              </div>
              <div
                aria-checked="false"
                aria-selected="false"
                class="item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  Sheep
                </span>
              </div>
              <div
                aria-checked="false"
                aria-selected="false"
                class="item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  Alpaca
                </span>
              </div>
              <div
                aria-checked="false"
                aria-selected="false"
                class="item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  Ass
                </span>
              </div>
              <div
                aria-checked="false"
                aria-selected="false"
                class="item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  Goat
                </span>
              </div>
              <div
                aria-checked="false"
                aria-selected="false"
                class="item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  Llama
                </span>
              </div>
              <div
                aria-checked="false"
                aria-selected="false"
                class="item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  Mule
                </span>
              </div>
            </div>
          </div>
        </div>
      </td>
      <td
        class="collapsing"
      >
        <div
          class="field"
        >
          <div
            class="ui fluid input"
          >
            <input
              aria-label="livestock count"
              id="field_input_schedule.grazingScheduleEntries.0.livestockCount"
              name="schedule.grazingScheduleEntries.0.livestockCount"
              type="text"
              value="10"
            />
          </div>
        </div>
      </td>
      <td
        class="collapsing"
      >
        <div
          class="field calendar-input--no-icon"
        >
          <div
            class="ui fluid icon input"
          >
            <i
              aria-hidden="true"
              class="icon"
            />
            <input
              aria-label="date in"
              name="schedule.grazingScheduleEntries.0.dateIn"
              tabindex="0"
              type="text"
              value="Nov 13"
            />
          </div>
        </div>
      </td>
      <td
        class="collapsing"
      >
        <div
          class="field calendar-input--no-icon"
        >
          <div
            class="ui fluid icon input"
          >
            <i
              aria-hidden="true"
              class="icon"
            />
            <input
              aria-label="date out"
              name="schedule.grazingScheduleEntries.0.dateOut"
              tabindex="0"
              type="text"
              value="Nov 13"
            />
          </div>
        </div>
      </td>
      <td
        class="collapsing"
      >
        1
      </td>
      <td
        class="collapsing"
      >
        <div
          class="field"
        >
          <div
            class="ui fluid transparent input"
          >
            <input
              type="text"
              value="4"
            />
          </div>
        </div>
      </td>
      <td
        class="collapsing"
      >
        0
      </td>
      <td
        class="collapsing"
      >
        0
      </td>
      <td
        class="collapsing center aligned"
      >
        <div
          aria-expanded="false"
          class="ui right pointing dropdown"
          role="listbox"
          tabindex="0"
        >
          <i
            aria-hidden="true"
            class="ellipsis vertical icon"
          />
          <div
            class="menu transition"
          >
            <div
              aria-checked="false"
              aria-selected="true"
              class="selected item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                Duplicate
              </span>
            </div>
            <div
              aria-checked="false"
              aria-disabled="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                Delete
              </span>
            </div>
          </div>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
